### PR TITLE
MM-54486 : Copy pasting images from chrome fails

### PR DESCRIPTION
**Pull request by M-ZubairAhmed**
#### Summary
**Cause**
An earlier pull request https://github.com/mattermost/mattermost/pull/23530, led to a regression issue. In this all clipboard events that contained HTML-type data. In typical clipboard events were blocked, there can be multiple types of data present. In this particular case, if an event contained both an image and HTML data, it was causing the blocking of pasting the image.

**Solution**
The paste handler for file uploads is specifically made to handle clipboard events that contain files. All other types of clipboard events are managed by different handlers. Therefore, if a clipboard event contains the following:
- Image + Text 🧽 Text 
- Image 🧽 Image
- Text 🧽 Text

#### Ticket Link
https://mattermost.atlassian.net/browse/MM-54486

#### Screenshots

https://github.com/mattermost/mattermost/assets/17708702/a91edf4f-ba63-42a9-9e81-60196e14b87f

#### Release Note

```release-note
NONE
```

### CODE REVIEW

File: webapp/channels/src/utils/paste.tsx [hunk 1]
```diff
@@ -42,6 +43,20 @@ export function isTextUrl(clipboardData: DataTransfer): boolean {
...
+/**
+ * Checks if the clipboard data contains plain text from list of types.
+**/
+export function hasPlainText(clipboardData: DataTransfer): boolean {
+    if (Array.from(clipboardData.types).includes('text/plain')) {
+        const clipboardText = clipboardData.getData('text/plain');
+        if (clipboardText.trim().length > 0) {
+            return true;
+        }
+        return false;
```
Comment (marianunez) [inline webapp/channels/src/utils/paste.tsx:55]: minor nit 

```suggestion
      return clipboardText.trim().length > 0;
```

File: webapp/channels/src/components/file_upload/file_upload.tsx [hunk 2]
```diff
@@ -735,14 +709,44 @@ export class FileUpload extends PureComponent<Props, State> {
...
+        const month = now.getMonth() + 1;
+        const date = now.getDate();
+        const hour = now.getHours().toString().padStart(2, '0');
+        const minute = now.getMinutes().toString().padStart(2, '0');
+
+        name = `${formatMessage(holders.pasted)}${year}-${month}-${date} ${hour}-${minute}${ext}`;
+    }
+
+    return new File([file as Blob], name, {type: file.type});
+}
```
Comment (marianunez) [inline webapp/channels/src/components/file_upload/file_upload.tsx:748]: Much easier to read with this extracted function 💯

**Review by marianunez — APPROVED** [review]
Thanks @M-ZubairAhmed! LGTM - great tests 🎉

File: webapp/channels/src/utils/paste.tsx (hunk 1 above)
Comment (M-ZubairAhmed) [reply webapp/channels/src/utils/paste.tsx]: Done ✅

**Comment by hmhealey** [comment]
I tested through a ton of different combinations of copying and pasting between Firefox, Chrome, Safari, and Desktop on Mac and Edge on Windows 11, and the only cases I found where it didn't work 100% were:
- Non-PNG images copied from Firefox, Chrome, and Edge were converted to PNGs. This seems like existing behaviour since it also happens on Slack.
- Images copied from Safari are pasted as text. In Safari, the image URL or linked URL is pasted, and in other browsers, a Markdown image linking to that URL is pasted. This doesn't happen on Slack, so we should probably fix this.
- Copying multiple files from Finder and pasting them into Firefox causes only the first file to be uploaded. This matches the current behaviour and Slack, so it's likely a browser limitation.

I also confirmed that copying and pasting plain text from OneNote still works as expected

File: webapp/channels/src/components/file_upload/file_upload.tsx (hunk 2 above)
Comment (hmhealey) [reply webapp/channels/src/components/file_upload/file_upload.tsx]: Agreed

**Review by hmhealey — APPROVED** [review]
I spun the Safari bug I mentioned above into a separate ticket (https://mattermost.atlassian.net/browse/MM-54864) since it's not a regression from what I can tell.

@yasserfaraazkhan If you have a good way to test if it's a regression or not, let us know, but we can probably open it as help wanted if it's not one

**Comment by yasserfaraazkhan** [comment]
@M-ZubairAhmed , @hmhealey , 
Copying image on safari seems to be regression.
I tried it on  https://test-images.test.mattermost.cloud 8.0.0.
The whole image gets uploaded and not the link as we see on 9.1

I tried the changes on Mac, copying Only images ,
- from Mac local to -> chrome, Firefox , safari browse works ✅ 
- from Win local -> MS edge ✅ 
- from Win online DOCX, EXECL  -> MS edge ❕ not working.

Also, as mentioned in QA steps the expected behaviour of pasting combination of TEXT + Image is not working on Mac/chrome

https://github.com/mattermost/mattermost/assets/15979783/54eff365-e1ee-4c37-a5b0-11213b39ebcb

**Comment by hmhealey** [comment]
>  Copying image on safari seems to be regression.

That's interesting since it's a bug in Slack as well I guess it's a regression since 8.0, but not on this PR, so I think we could still leave it until that other ticket then. We definitely don't want to break those other cases though

**Review by yasserfaraazkhan — APPROVED** [review]
Hi @M-ZubairAhmed the test and images are copied separately.
from chrome/ Firefox/ local to MM web and desktop app

**Comment by M-ZubairAhmed** [comment]
| Word | One Note | Excel | Webpage
-- | -- | -- | -- | --
MacOS Chrome | T/P/T | T/P/T | T/P/M | T/P/M
MacOS Firefox | T/P/T | T/P/T | T/P/M | T/P/M
MacOS Edge | T/P/T | T/P/T | T/P/M | T/P/M
MacOS Safari | T/P/T | T/P/T | T/P/M | T/T/M
MacOS Electron |   |   |   |  
Windows Chrome |   |   |   |  
Windows Firefox |   |   |   |  
Windows Edge |   |   |   |  
Windows Electron |   |   |   |  
Ubuntu Chrome |   |   |   |  
Ubuntu Firefox |   |   |   |  
Ubuntu Edge |   |   |   |  
Ubuntu Electron |   |   |   |  

 


#### Legend
Flow - Text copied / Picture copied / Text and Picture copied
T - Text pasted
P - Picture pasted
M - Markdown pasted

**Comment by amyblais** [comment]
@M-ZubairAhmed Please help open a manual cherry-pick PR to backport this to ``release-8.1`` branch.

**Pull request merged**